crates/autopilot/src/domain/fee/mod.rs

```diff
@@ -185,7 +186,27 @@ impl ProtocolFees {
             })
             .flat_map(|policy| Self::variant_fee_apply(&order, &quote, policy))
             .chain(partner_fees)
+            // Volume fees can only be taken by waiting for the market price to
+            // become sufficienly better than the order's limit price that we can
+            // take the volume fee cut without violating the order's limit price
+            // afterwards.
+            // That means if an order has a volume fee and a surplus fee and we
+            // were to take a cut for the surplus fee first it would be unreasonably
+            // high while the cut taken for the volume fee woule be unreasonbly low.
+            // That's why we want to apply volume fees first (while preserving a higher
+            // priority for protocol fees over partner fees). This is achieved by
+            // stable sorting volume fees to the front of the vector.
+            .sorted_by(|a, b| {
+                // function sorts in ascending order so items with priority 0 will
+                // appear before items with priority 1.
+                let prio = |policy: &Policy| match policy {
```

Comment:
nit: if we want the policy priority to be more explicit and reusable, we could implementing `Ord` and `PartialOrd` traits for `Policy`.